### PR TITLE
[FLINK-1962] Add Gelly Scala API v2

### PR DESCRIPTION
**Pull request by PieterJanVanAeken**
Second version of the PR for the Scala Gelly API. For more information, see PR #808

### CODE REVIEW

**Comment by StephanEwen** [comment]
Wow, looks like a good piece of work. Nicely with tests and everything.
Build also passes, style looks good.

+1 to merge this from my side. I'd like to wait for a day or two to get a comment from one of the Gelly people (Vasia or Andra).

One thing, though: In the Batch and streaming APIs, we added a "completeness check" to make sure that methods added to the Java APIs are also present in the Scala APIs. Would that be a good thing to add here as well?

**Comment by PieterJanVanAeken** [comment]
I would postpone adding the "completeness check" as it will currently fail. Since I started working on this, the Java Gelly API has changed and while I modified my work to be compatible with the changes, not all new Java Gelly methods have a Scala counterpart yet. It was discussed briefly in the initial PR #808 and it was decided that adding these new methods should go into a separate JIRA issue.

**Comment by StephanEwen** [comment]
Sounds good.

If no one objects in the next days, I would merge this.

**Comment by vasia** [comment]
+1 from me too. Thanks for your great work @PieterJanVanAeken!

File: flink-staging/flink-gelly-scala/src/main/scala/org/apache/flink/graph/scala/Graph.scala [hunk 1]
```diff
@@ -0,0 +1,735 @@
...
+   * @return the resulting sub-graph.
+   */
+  def filterOnEdges(edgeFilterFun: Edge[K, EV] => Boolean) = {
+    val edgeFilter = new FilterFunction[Edge[K, EV]] {
+      val cleanEdgeFun = clean(edgeFilterFun)
+
+      override def filter(value: Edge[K, EV]): Boolean = cleanEdgeFun(value)
+    }
+
+    //wrapGraph(jgraph.filterOnEdges(edgeFilter))
```
Comment (chiwanpark) [inline flink-staging/flink-gelly-scala/src/main/scala/org/apache/flink/graph/scala/Graph.scala]: Maybe unnecessary comment?

**Comment by chiwanpark** [comment]
Looks good to merge except one cosmetic issue. Awesome job! :)

**Comment by PieterJanVanAeken** [comment]
I fixed the unneccessary comment and did a quick rebase to keep up with master branch.

**Comment by StephanEwen** [comment]
All right, let's merge this!

**Comment by rmetzger** [comment]
What are the plans for documenting the Gelly Scala API ?

**Comment by StephanEwen** [comment]
If the APIs are in sync there should be one documentation for both, right?
Similar as with the other APIs...

**Comment by PieterJanVanAeken** [comment]
As mentioned, they are not in sync yet. We need to create an issue to update the Scala API so that it is in sync with the Java API. A completeness test should also be added once that is done. 

Since this API just a wrapper, no additional documentation seems needed but we could add a "java/scala" tab to the documentation, similar to the Batch API documentation. It would look extremely similar though, so I'm not sure how much added value it would have.

**Comment by vasia** [comment]
I have created [FLINK-2561](https://issues.apache.org/jira/browse/FLINK-2561) for the sync.

**Pull request closed without merging**